calour/export_html.py

```diff
@@ -0,0 +1,170 @@
+    if cmap is None:
+        cmap = plt.rcParams['image.cmap']
+    # plot the heatmap with 1 pixel per feature/sample, no axes/lines
+    fig = plt.figure(frameon=False, dpi=300)
+    fig.set_size_inches(exp.shape[0]/300, exp.shape[1]/300)
+    ax = plt.Axes(fig, [0., 0., 1., 1.])
+    ax.set_axis_off()
+    fig.add_axes(ax)
+    ax.imshow(data.transpose(), interpolation='nearest')
+
+    if title is None:
+        title = exp.description
+
+    # add parameters to html template
+    with open(html_template) as fl:
+        html_page = fl.read()
+    html_page = html_page.replace('// yticklabels go here', 'yticklabels=%s' %
+                                  _list_to_string(exp.feature_metadata[feature_field].values))
+    html_page = html_page.replace('// ids go here', 'ids=%s' %
+                                  _list_to_string(exp.feature_metadata.index.values))
+    html_page = html_page.replace('// samples go here', 'samples=%s' %
+                                  _list_to_string(exp.sample_metadata.index.values))
+    if sample_field is not None:
+        html_page = html_page.replace('// field_name goes here', 'field_name = "%s"' %
+                                      sample_field)
+    html_page = html_page.replace('// title_text goes here', 'title_text="%s"' %
+                                  title)
+
+    # add vertical lines between sample groups and add x tick labels
+    if sample_field is not None:
+        try:
+            xticks = _transition_index(exp.sample_metadata[sample_field])
+        except KeyError:
+            raise ValueError('Sample field %r not in sample metadata' %
+                             sample_field)
+        x_pos, x_val = zip(*xticks)
+        x_pos = np.array([0.] + list(x_pos))
+
+        html_page = html_page.replace('// vlines go here', 'vlines=%s' %
+                                      _list_to_string(x_pos[1:-1]))
+        xtick_pos = x_pos[:-1] + (x_pos[1:] - x_pos[:-1]) / 2
+        html_page = html_page.replace('// xtick_pos go here', 'xtick_pos=%s' %
+                                      _list_to_string(xtick_pos))
+
+        xticklabels = [str(i) for i in x_val]
+        # shorten x tick labels that are too long:
+        if xticklabel_len is not None:
+            mid = int(xticklabel_len / 2)
+            xticklabels = ['%s..%s' % (i[:mid], i[-mid:])
+                           if len(i) > xticklabel_len else i
+                           for i in xticklabels]
+        html_page = html_page.replace('// xtick_labels go here', 'xtick_labels=%s' %
+                                      _list_to_string(xticklabels))
+
+    # embed the figure png into the html page
+    figfile = BytesIO()
+    fig.savefig(figfile, format='png', dpi=300)
+    figfile.seek(0)  # rewind to beginning of file
+    import base64
+    figdata_png = base64.b64encode(figfile.getvalue())
+    figfile.close()
+    html_page = html_page.replace('**image_goes_here**', urllib.parse.quote(figdata_png))
```

Comment:
mv `urllib.parse.quote` to `with` block. just use the final value in this `replace` call